net: http: Add compression support in HTTP server

kica-z opened the pull request:
Add compression support using the accept-encoding header to the http server static filesystem resource.

In our use case we have different files with different compressions it is therefore important that the webserver is able to handle those compressions correctly. This is a first draft of the feature. I am open to discussion on how to implement this or if this is even wanted in upstream.

jukkar reviewed (commented):
No objection to the feature, just wondering if it would make sense to have this more configurable. I mean that if user only needs gzip support, then the other methods are useless to compile in. Perhaps have a Kconfig entries for each supported compression method, and then compile out code that is not needed.

kica-z commented:
@jukkar I agree in the sense that a kconfig to enable compression in general would be good. I am however not sure if kconfigs for each compression method would't be to much as the only thing they would remove is one `if`  in the  `http_server_find_file` function. What do you think? I added a general compression kconfig for now.

I will also have a look at the tests that are failing (need to add tests covering compression anyhow).

rlubos reviewed (commented):
Any chance for some test for this? Or at least we should enable it in the sample to make sure it builds.

jukkar commented on include/zephyr/net/http/compression.h:
```diff
@@ -0,0 +1,54 @@
+	HTTP_ZSTD = 4      /**< ZSTD */
+};
+
+void http_compression_parse_accept_encoding(const char *accept_encoding, size_t len,
```
And also the functions should be documented using doxygen comments.

kica-z replied:
@rlubos My thought was that it would be nice for developers to be able to also get the supported compressions from the client context when needed.

Or are you just talking about the function `http_compression_parse_accept_encoding`?

jukkar commented:
> What do you think? I added a general compression kconfig for now.

What you have now (one Kconfig option) looks fine.

kica-z commented:
> Any chance for some test for this? Or at least we should enable it in the sample to make sure it builds.

Yes I will add some tests for this 👍

kica-z commented on subsys/net/lib/http/http_compression.c:
```diff
@@ -0,0 +1,129 @@
+	return "";
+}
+
+int http_compression_from_text(enum http_compression compression[static 1], const char *text)
```
`[static 1]` is supported by the c standard since C99 i believe. It's purpose is to tell the developer as well as the compiler that this argument does not accept `NULL` as value. This can then also be checked by enabling the respective compiler check.

I will remove this, it is primarily in there because we use it in our codebase.

kica-z commented on subsys/net/lib/http/http_server_core.c:
```diff
@@ -785,26 +787,67 @@ struct http_resource_detail *get_resource_detail(const struct http_service_desc
-	if (ret < 0) {
-		len = strlen(fname);
+	len = strlen(fname);
+#ifdef CONFIG_HTTP_SERVER_COMPRESSION
```
I pushed a new version with more ifdefs but as much as possible `if(IS_ENABLED`

rlubos replied:
Sure, `if (IS_ENALBED())` cannot be used in all cases, but IMHO the current approach already improved code readability, I think we're good.

rlubos commented on subsys/net/lib/http/headers/server_internal.h:
```diff
@@ -36,13 +36,22 @@ int enter_http1_request(struct http_client_ctx *client);
+/* HTTP Compression handling */
+#define HTTP_COMPRESSION_MAX_STRING_LEN 8
+void http_compression_parse_accept_encoding(const char *accept_encoding, size_t len,
+	uint8_t *supported_compression);
```
Please align to the opening bracket.

rlubos commented on subsys/net/lib/http/http_server_http1.c:
```diff
@@ -506,7 +512,12 @@ int handle_http1_static_fs_resource(struct http_resource_detail_static_fs *stati
 	/* open file, if it exists */
-	ret = http_server_find_file(fname, sizeof(fname), &file_size, &gzipped);
+#ifdef CONFIG_HTTP_SERVER_COMPRESSION
+ret = http_server_find_file(fname, sizeof(fname), &file_size, client->supported_compression,
+			    &chosen_compression);
```
Indentation is missing.

rlubos reviewed (commented):
Please also check compliance reports - there are some issues here and there. Generally looks good to me - would still appreciate those tests though.

kica-z commented:
@rlubos I fixed the compliance problems. I will get to the tests as soon as i find the time.

kica-z commented:
@rlubos I added tests regarding compression support.

mrodgers-witekio commented on subsys/net/lib/http/http_compression.c:
```diff
@@ -0,0 +1,133 @@
+				  enum http_compression expected)
+{
+	assert(compression);
+	assert(text);
```
Probably better to use `__ASSERT` or `__ASSERT_NO_MSG` from `zephyr/sys/_assert.h` here, I'm not sure if the normal C assert is correctly enabled/disabled by `CONFIG_ASSERT` in Zephyr, or if it'll have the same error message formatting?

kica-z committed:
net: http: Add compression support in HTTP server

Add compression support using the accept-encoding
header to the http server static filesystem resource.

Signed-off-by: Carlo Kirchmeier <carlo.kirchmeier@zuehlke.com>

kica-z committed:
test: net: http: Add compression support in HTTP server

Add tests verifying the functionality of the newly added
accept-encoding based compression support.

Signed-off-by: Carlo Kirchmeier <carlo.kirchmeier@zuehlke.com>
